flipper/contrib/postgresql.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2018 eShares, Inc. dba Carta, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+from contextlib import contextmanager
+from typing import Iterator, Optional
+
+from psycopg import Connection, connect, sql
+
+from .interface import AbstractFeatureFlagStore, FlagDoesNotExistError
+from .storage import FeatureFlagStoreItem, FeatureFlagStoreMeta
+from .util.date import now
+
+
+class PostgreSQLFeatureFlagStore(AbstractFeatureFlagStore):
+    def __init__(
+        self,
+        conninfo: str,
+        table_name: str = "feature_flags",
+    ) -> None:
+        self._conninfo = conninfo
+        self._table_name = sql.Identifier(table_name)
+        self._name_column = sql.Identifier("name")
+        self._item_column = sql.Identifier("item")
+        self._run_migrations()
+
+    @contextmanager
+    def _connection(self) -> Connection:
+        conn = connect(self._conninfo)
+        try:
+            yield conn
+        finally:
+            conn.close()
+
+    def _run_migrations(self) -> None:
+        with self._connection() as conn:
+            query = sql.SQL(
+                """CREATE TABLE IF NOT EXISTS {} \
+                ({} varchar(40) PRIMARY KEY, {} bytea NOT NULL)"""
+            ).format(self._table_name, self._name_column, self._item_column)
+            conn.execute(query)
+            conn.commit()
+
+    def _update(self, item: FeatureFlagStoreItem) -> None:
+        with self._connection() as conn:
+            query = sql.SQL("UPDATE {} SET {} = %s WHERE {} = %s").format(
+                self._table_name, self._item_column, self._name_column
+            )
+            conn.execute(query, (item.serialize(), item.feature_name))
+            conn.commit()
+
+    def create(
+        self,
+        feature_name: str,
+        is_enabled: bool = False,
+        client_data: Optional[dict] = None,
+    ) -> FeatureFlagStoreItem:
+        item = FeatureFlagStoreItem(
+            feature_name, is_enabled, FeatureFlagStoreMeta(now(), client_data)
+        )
+        if self.get(feature_name) is None:
+            with self._connection() as conn:
+                query = sql.SQL("INSERT INTO {} ({}, {}) VALUES (%s, %s)").format(
+                    self._table_name, self._name_column, self._item_column
+                )
+                conn.execute(query, (item.feature_name, item.serialize()))
+                conn.commit()
+        else:
+            self._update(item)
```

Comment:
How would you feel about condensing this down into a single query? Since you're using the feature name as the pk, there should automatically be a unique constraint. This would allow us to use `INSERT ... ON CONFLICT DO NOTHING` semantics.